README.md
# Savvy Student Stats :sunglasses:

<details>
<summary>Contents</summary>

## Contents

- [Awesome School Stats :sunglasses: (`school-stats-powerbi`)](#awesome-school-stats-sunglasses-school-stats-powerbi)
  - [Contents](#contents)
  - [Intro](#intro)
  - [HOWTO Obtain Security Key for Moodle](#howto-obtain-security-key-for-moodle)
  - [History](#history)
    - [TL;DR (somewhat)](#tldr-somewhat)
  - [More History](#more-history)

</details>

----
## Intro

Using the Moodle web service, this package extracts Course and Grade info to feed into Power BI for pretty reports and AI insights!

Hope it helps someone. If not, I at least have a copy of the repo for my own purposes. 
Since my son's education is the most important thing, after managing Type-1 Diabetes, is school.
It is learning how to learn, develop ing productive habits, and handling feedback.
It determines the future.

> "Knowledge is power"
> --- *Sir Francis Bacon (or Thomas Jefferson, apparently there's flame wars)*

---

## HOWTO Obtain Security Key for Moodle

1. Login to `Moodle` as the Student.
2. Click the name/avatar at the top right.
3. Click `Preferences`.
4. Click `Security keys`.
5. Should end up on a URL ending 
   with something like 
   `/user/managetoken.php?sesskey=YOUR_SESSION_ID_THAT_DOES_NOT_MATTER`.





![image-Moodle-Preferences](README.assets/image-Moodle-Preferences.png)

![image-Moodle-Keys](README.assets/image-Moodle-Keys.png)



---


## History

### TL;DR (somewhat)

This started as a summary but turned into the full description. Hope it's useful.

My 13 year old son ("Bob") switched to online school to finish Grade 6 in 2020. We
decided to stay in online school for Grade 7. Tried using Notion and other
management techniques to track assignments. Process was tedious and manual.
Wound up in a two week sprint of getting things together during Jan/Feb 2021.

During this, as a parent, I wanted to see what assignments had not been handed
in, and what subjects my son was doing the best/worst in.

I just started using `Cypress` for automated browser testing and saw a chance to play with it,
push it outside the usual testing use case and turn it into a web scraper.

I had access to the "My Mentees" grade report through the school's 
online learning management system (LMS), which for Grade 6 was Google Classroom,
and for Grade 7 (Bob's current grade) the school used Moodle.

Moodle feels older, more limited, than Google Classroom.
I never got as deep into stats, analytics, devops, etc.. when he was using Google Classroom.

So I wrote some Cypress tests that would:

1. Login to Moodle
2. Visit the `My Mentee Grades Report`
3. Scrape the assignment feedback from the report.

The problems I found were:

1. Only some assignments were listed
   1. Teachers can hide assignments.
   2. I think that some assignments are auto-hidden after a certain time.
   3. I believe from past conversations that teachers like to keep the report page clean, for both the `Student` and the `Mentor (Parent)`.
   4. Different subjects have different report pages.
      1. Later found out that is because Moodle has a "`Grading Book`" which seems to be configured per `Course` *(subject)* 
         1. Is "subject" a local term? I'm from Calgary, Alberta, Canada 🍁.
   5. Final nail in the coffin was that I couldn't extract any date for assignments. 
      1. No idea what was sold and new.



## More History

As a parent of an Online Grade 7 Jr. High Student in Canada, I am helping my
son daily, hourly, and sometimes every 10 minutes-ish, to help him:

- Decide what to focus on, well learning how to manage himself
- Review school assignments
- Learn the subject matter in detail
- Study for quizes
- Manage Assignments
- Using a Kanban Board in [Notion](//notion.so)
- Follow-up on submitted assignments
  - Check for Feedback (grade and comments)
  - Reply to any actionable feedback
    - Common feedback at the beginning of the year was that a file would not
      be attached properly (sometimes due to confusing language on the
      submission page), or a few questions would be overlooked that needed
      responses added.
- Reflect on Feedback
  - What was awesome?
  - What can be done better?
  - How difficult was it?
  - Is the grade going to achieve goals for the Reporting Period, week, month,
    etc.

After close to six months online finishing Grade 6,(Feb-July 2020), when
COVID-19 became a global pandemic, me, my son, and his mom, decided to enroll
in online school for Grade 7 (Sep-July 2020/2021). 

There were numerous benefis to online scool, including:

1. FAR less sick days.
   - By far the number one benefit to us is that as a child with Type-1
     Diabetes, a common cold easily becomes a major infection, knocking him
     out for a few days, or a week at a time. 
   - There were over 20 sick days in Grade 6, and countless days not feeling
     well but still attending.
   - 

TODO: Finish


---
Installation
Usage
Contributing
Credits
License


---

*Made with :heart: by [sqrl (Dad)](https://twitter.com/gregdevelops) and [Bobthebe3t (Child/Student)](http://www.bobisagamer.com/)*.

---


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog][Keep a Changelog] and this project (somewhat) adheres to [Semantic Versioning][Semantic Versioning].

The [changelog-generator vscode extension][changelog-generator]  was used to create this file.
Trying `ember` and other templates. Whatever suits my needs, tickles my fancy.The vscode extension may, or may-not, be used in future, but it is very neat.


## [Unreleased]


---

## [Released]

## 2022-06-02 - v0.14.2 (?) - Cleanup in case of traffic coming from `TRXN8R`
- `TRXN8R | Excel Power Query Bulk CSV Financial Transactions Importer`
- https://github.com/trxn8r/trxn8r
- Quickly Edited Name as this may get some traffic thanks to new CSV Importer/Transformer (Power Query ETL)
github.com/trxn8r/trxn8r
- Surprised this was 18 months ago, seems like both a longer and shorter time.

### 2021-02-14 - v0.14 - Power BI Report Works with Courses, Grades, and Assignments!
Skipping 0.13 b/c lots of changes and 13 is said to be unlucky.

Got it working last night around 2 AM.
Pubished to Power BI Workspace.
Was so freaking excited!
Still am :)

It's everything I could have hoped for, and more!

Not sure about ths changelog title format, probably change in future Trying to fcus on actually finishing the Power BI report, rather than document everything. Hmm...

---

<!-- Links -->
[Keep a Changelog]: https://keepachangelog.com/
[Semantic Versioning]: https://semver.org/
[changelog-generator]: https://marketplace.visualstudio.com/items?itemName=axetroy.vscode-changelog-generator


<!-- Versions -->
[Unreleased]: https://github.com/Author/Repository/compare/v1.0.0...HEAD
[Released]: https://github.com/Author/Repository/releases
[0.0.2]: https://github.com/Author/Repository/compare/v0.0.1..v0.0.2
[0.0.1]: https://github.com/Author/Repository/releases/v0.0.1


powerbi-snippets/source_gradereport_user_get_grade_items.alternate-using-course-table.m
// zzz-source-gradereport_user_get_grade_items
// @since 2021-02-10

let
    Source = #"Source-get_users_courses",
    #"Removed Other Columns" = Table.SelectColumns(Source,{"Course.id", "Course.shortname"}),

    RemoteData = AskMoodleFor(
        "gradereport_user_get_grade_items",
        "userid=" & MOODLE_USER_ID
        & "&courseid=" [Course.id]
    ),
    #"Invoked Custom Function" = Table.AddColumn(
        #"Removed Other Columns", "GradeReport",
        each AskMoodleFor("gradereport_user_get_grade_items", "userid="REPLACE"&courseid=REPLACE")
    )
in
    #"Invoked Custom Function"


---

// Source-Gradebooks.dax
// @since 2021-0

let
    COURSE_IDS = { 780,671,673,684,702 },

    // thanks `#table` https://blog.crossjoin.co.uk/2016/06/03/creating-tables-in-power-bipower-query-m-code-using-table/
    SourceTable = #table({"G.courseId"}, { {780},{671},{673},{684},{702} }),
    #"Invoked Custom Function" = Table.AddColumn(SourceTable, "AskMoodleForGradebook", each AskMoodleForGradebookTable([courseid])),
    #"Expanded AskMoodleForGradebook" = Table.ExpandRecordColumn(#"Invoked Custom Function", "AskMoodleForGradebook", {"usergrades"}, {"usergrades"}),
    #"Expanded usergrades" = Table.ExpandListColumn(#"Expanded AskMoodleForGradebook", "usergrades"),
    #"Expanded usergrades1" = Table.ExpandRecordColumn(#"Expanded usergrades", "usergrades", {"courseid", "userid", "userfullname", "maxdepth", "gradeitems"}, {"courseid.1", "userid", "userfullname", "maxdepth", "gradeitems"}),
    #"Expanded gradeitems" = Table.ExpandListColumn(#"Expanded usergrades1", "gradeitems"),
    #"Expanded gradeitems1" = Table.ExpandRecordColumn(#"Expanded gradeitems", "gradeitems", {"id", "itemname", "itemtype", "itemmodule", "iteminstance", "itemnumber", "categoryid", "outcomeid", "scaleid", "locked", "graderaw", "gradedatesubmitted", "gradedategraded", "gradehiddenbydate", "gradeneedsupdate", "gradeishidden", "gradeislocked", "gradeisoverridden", "gradeformatted", "feedback", "feedbackformat", "cmid"}, {"gradeitems.id", "gradeitems.itemname", "gradeitems.itemtype", "gradeitems.itemmodule", "gradeitems.iteminstance", "gradeitems.itemnumber", "gradeitems.categoryid", "gradeitems.outcomeid", "gradeitems.scaleid", "gradeitems.locked", "gradeitems.graderaw", "gradeitems.gradedatesubmitted", "gradeitems.gradedategraded", "gradeitems.gradehiddenbydate", "gradeitems.gradeneedsupdate", "gradeitems.gradeishidden", "gradeitems.gradeislocked", "gradeitems.gradeisoverridden", "gradeitems.gradeformatted", "gradeitems.feedback", "gradeitems.feedbackformat", "gradeitems.cmid"}),
    #"Removed Other Columns" = Table.SelectColumns(#"Expanded gradeitems1",{"gradeitems.id", "gradeitems.itemname", "gradeitems.itemtype", "gradeitems.itemmodule", "gradeitems.iteminstance", "gradeitems.itemnumber", "gradeitems.categoryid", "gradeitems.outcomeid", "gradeitems.scaleid", "gradeitems.locked", "gradeitems.graderaw", "gradeitems.gradedatesubmitted", "gradeitems.gradedategraded", "gradeitems.gradehiddenbydate", "gradeitems.gradeneedsupdate", "gradeitems.gradeishidden", "gradeitems.gradeislocked", "gradeitems.gradeisoverridden", "gradeitems.gradeformatted", "gradeitems.feedback", "gradeitems.feedbackformat", "gradeitems.cmid", "courseid"})
in
    #"Removed Other Columns"


powerbi-src/zzz-source-gradebooks-raw.m
// zzz-source-gradereport_user_get_grade_items
// @since 2021-02-10
let

    // thanks `#table` https://blog.crossjoin.co.uk/2016/06/03/creating-tables-in-power-bipower-query-m-code-using-table/
    // COURSE_IDS = { 780,671,673,684,702 },
    NewTable = #table(
        {"GI.courseId"},
        {
            {780},
            {671},
            {673},
            {684},
            {702}
        }
    ),

    #"Get Gradebooks from Moodle" = Table.AddColumn(
        NewTable,
        "MoodleResponse",
        each AskMoodleForGradebookByCourse([GI.courseId])
    ),


    #"Expanded AskMoodleForGradebook" = Table.ExpandRecordColumn(
        #"Get Gradebooks from Moodle",
        "MoodleResponse",
        {"usergrades"},
        {"usergrades"}
    ),
    #"Expanded usergrades" = Table.ExpandListColumn(
        #"Expanded AskMoodleForGradebook",
        "usergrades"
    ),
    #"Expanded usergrades1" = Table.ExpandRecordColumn(
        #"Expanded usergrades",
        "usergrades",
        {
            // "courseid",
            // "userid",
            // "userfullname",
            // "maxdepth",
            "gradeitems"
        },
        {
            // "usergrades.courseid",
            // "usergrades.userid",
            // "userfullname",
            // "maxdepth",
            "gradeitems"
        }
    ),

    #"Expanded gradeitems" = Table.ExpandListColumn(
        #"Expanded usergrades1",
        "gradeitems"
    ),

    #"Expanded gradeitems1" = Table.ExpandRecordColumn(
        #"Expanded gradeitems",
        "gradeitems",
        {
            "id",
            "itemname",
            "itemtype",
            "itemmodule",
            "iteminstance",
            "itemnumber",
            "categoryid",
            "outcomeid",
            "scaleid",
            "locked",
            "graderaw",
            "gradedatesubmitted",
            "gradedategraded",
            "gradehiddenbydate",
            "gradeneedsupdate",
            "gradeishidden",
            "gradeislocked",
            "gradeisoverridden",
            "gradeformatted",
            "feedback",
            "feedbackformat",
            "cmid"
        },
        {
            "GI.id",
            "GI.itemname",
            "GI.itemtype",
            "GI.itemmodule",
            "GI.iteminstance",
            "GI.itemnumber",
            "GI.categoryid",
            "GI.outcomeid",
            "GI.scaleid",
            "GI.locked",
            "GI.graderaw",
            "GI.gradedatesubmitted",
            "GI.gradedategraded",
            "GI.gradehiddenbydate",
            "GI.gradeneedsupdate",
            "GI.gradeishidden",
            "GI.gradeislocked",
            "GI.gradeisoverridden",
            "GI.gradeformatted",
            "GI.feedback",
            "GI.feedbackformat",
            "GI.cmid"
         }
    ),
    #"Removed Other Columns" = Table.SelectColumns(
        #"Expanded gradeitems1",
        {
            "GI.id",
            "GI.itemname",
            "GI.itemtype",
            "GI.itemmodule",
            "GI.iteminstance",
            "GI.itemnumber",
            "GI.categoryid",
            "GI.outcomeid",
            "GI.scaleid",
            "GI.locked",
            "GI.graderaw",
            "GI.gradedatesubmitted",
            "GI.gradedategraded",
            "GI.gradehiddenbydate",
            "GI.gradeneedsupdate",
            "GI.gradeishidden",
            "GI.gradeislocked",
            "GI.gradeisoverridden",
            "GI.gradeformatted",
            "GI.feedback",
            "GI.feedbackformat",
            "GI.cmid"
        }
    )

    // ADD DATE COLUMNS

in
    #"Removed Other Columns"

